ApiReview/API.Polly.Core/NoDocs/Polly/OutcomeArguments.cs

```diff
@@ -0,0 +1,16 @@
+// Assembly 'Polly.Core'
+
+using System;
+using System.Runtime.CompilerServices;
+
+namespace Polly;
+
+public readonly struct OutcomeArguments<TResult, TArgs>
```

Comment:
Can we drop `OutcomeArguments` and move `Context` and `Outcome` to each respective arguments?

For example:

``` csharp
public readonly struct OnRetryArguments<TResult>
{
    // moved here
    public Outcome<TResult> Outcome { get; }

    // moved here
    public ResilienceContext Context { get; }

    public int AttemptNumber { get; }

    public TimeSpan RetryDelay { get; }

    public TimeSpan ExecutionTime { get; }
}
```

Now the delegate signature instead of:

``` csharp
    public Func<OutcomeArguments<TResult, OnRetryArguments>, ValueTask>? OnRetry { get; set; }
```

Becomes:

``` csharp
    public Func<OnRetryArguments<TResult>, ValueTask>? OnRetry { get; set; }
```

Upsides:

- Cleaner and easier to consume API
- Arguments are aligned with non-outcome based ones (`OnTimeoutArguments` for example)

Downsides:

- Duplication of properties across multiple argument classes.
- Arguments are now generic which makes them harder to consume.

Reply:
Do you have an example of code consuming this for before and after this change? Would make it easier to make a judgement on the downsides for the consumer.

Reply:
Sure, for delegates it would look like:

``` csharp

// before
new RetryStrategyOptions
{
    OnRetry = args =>
    {
        Console.WriteLine("Attempt: {0}", args.Arguments.AttemptNumber);
        return default;
    }
}

// after
new RetryStrategyOptions
{
    OnRetry = args =>
    {
        Console.WriteLine("Attempt: {0}", args.AttemptNumber);
        return default;
    }
}
```

For telemetry:

``` csharp
//before
private class CustomTelemetryListener : TelemetryListener
{
    public override void Write<TResult, TArgs>(in TelemetryEventArguments<TResult, TArgs> args)
    {
        if (args.Arguments is OnRetryArguments retryArgs)
        {
            Console.WriteLine("Attempt: {0}", retryArgs.AttemptNumber);
        }
    }
}


// after
private class Listener : TelemetryListener
{
    public override void Write<TResult, TArgs>(in TelemetryEventArguments<TResult, TArgs> args)
    {
        if (args.Arguments is OnRetryArguments<TResult> retryArgs)
        {
            Console.WriteLine("Attempt: {0}", retryArgs.AttemptNumber);
        }
    }
}
```

Major use-case are the delegates where it looks simpler. 
For listeners it would be a problem before where the `Write` method wasn't generic. Now it's doable.

Reply:
I think the proposed change looks good then.

Reply:
@martintmk I guess not all the arguments are gonna be generic, only the ones for reactive strategies right? 

```c#
public readonly struct OnRetryArguments<TResult>
```

And yeah from the simplicity standpoint, I lean towards removing the `OutcomeArguments` envelope, actually, that's how I initially used it in Simmy for reactive chaos strategies.

Reply:
>  I guess not all the arguments are gonna be generic, only the ones for reactive strategies right?

Yes, reactive strategies only.

Thanks folks, I'll prepare the PR with changes.

Reply:
Addressed by #1513